--- lcommon	2020-12-30 15:25:51.308803281 +0000
+++ lcommon	2020-12-30 15:27:03.455103240 +0000
@@ -334,9 +334,15 @@
    if [ -z "$VHOSTLIST" ]; then
       echo 1>&2 "Warning: there are no virtual machines to be started."
    else
+      cur=$(bspc query -N -n focused)
+      desktop=$(bspc query -D -d 'next.!occupied')
+      bspc desktop -f $desktop
       for VM in $VHOSTLIST; do
+         bspc node -f $(bspc query -T -d $desktop | szhelper.rb)
          startSingleVM $VM
       done
+      sleep 2
+      bspc node -f $cur
    fi
    
 	# The readyfor.test file should not be created if we have been invoked
--- lstart	2020-12-30 15:25:35.435926358 +0000
+++ lstart	2020-12-30 15:51:44.146440140 +0000
@@ -370,7 +370,7 @@
 [ ! -z "$SCRIPTMODE" ] && echo -n "Starting lab... "
 # Check whether a lab.dep file exists. If so, enable parallel startup of
 # virtual machines.
-if [ -f "$LAB_DIRECTORY/lab.dep" -a -z "$BE_SEQUENTIAL" ] || [ ! -z "$FORCE_PARALLEL_MODE" ]; then
+if [ -f "$LAB_DIRECTORY/lab.dep" -a ! -z "$BE_SEQUENTIAL" ] || [ ! -z "$FORCE_PARALLEL_MODE" ]; then
    # Use parallel startup
    if [ -z "$SCRIPTMODE" ]; then
       becho "You chose to use parallel startup."
