diff -urpN b/arch/um/drivers/Makefile a/arch/um/drivers/Makefile
--- b/arch/um/drivers/Makefile	2013-12-16 19:28:47.782089822 +0100
+++ a/arch/um/drivers/Makefile	2013-12-16 19:30:49.150095750 +0100
@@ -17,9 +17,9 @@ ubd-objs := ubd_kern.o ubd_user.o
 port-objs := port_kern.o port_user.o
 harddog-objs := harddog_kern.o harddog_user.o
 
-LDFLAGS_pcap.o := -r $(shell $(CC) $(KBUILD_CFLAGS) -print-file-name=libpcap.a)
+LDFLAGS_pcap.o := -r -L$(shell dirname `$(CC) $(KBUILD_CFLAGS) -print-file-name=libpcap.a`) -lpcap
 
-LDFLAGS_vde.o := -r $(shell $(CC) $(CFLAGS) -print-file-name=libvdeplug.a)
+LDFLAGS_vde.o := -r -L$(shell dirname `$(CC) $(CFLAGS) -print-file-name=libvdeplug.a`) -lvdeplug
 
 targets := pcap_kern.o pcap_user.o vde_kern.o vde_user.o

diff -urpN b/arch/um/drivers/port_user.c a/arch/um/drivers/port_user.c
--- b/arch/um/drivers/port_user.c	2013-12-16 19:28:47.774089822 +0100
+++ a/arch/um/drivers/port_user.c	2013-12-16 19:30:49.174093915 +0100
@@ -5,6 +5,7 @@
 
 #include <stdio.h>
 #include <stdlib.h>
+#include <string.h>
 #include <errno.h>
 #include <termios.h>
 #include <unistd.h>
@@ -167,10 +168,31 @@
 int port_connection(int fd, int *socket, int *pid_out)
 {
 	int new, err;
-	char *argv[] = { "/usr/sbin/in.telnetd", "-L",
-			 OS_LIB_PATH "/uml/port-helper", NULL };
 	struct port_pre_exec_data data;
 
+	/************************************************/
+	/* Changed for better compatibility with Netkit */
+	/************************************************/
+
+/*	char *argv[] = { "/usr/sbin/in.telnetd", "-L",
+			 "/usr/lib/uml/port-helper", NULL }; */
+
+	char *netkit_home, port_helper_bin[4096];
+	
+	netkit_home = getenv("NETKIT_HOME");
+	if (!netkit_home)
+		netkit_home = getenv("VLAB_HOME");
+
+	if (!netkit_home)
+		strcpy (port_helper_bin, "/usr/lib/uml/port-helper");
+	else
+		sprintf (port_helper_bin, "%s/bin/port-helper", netkit_home);
+
+	char *argv[] = { "/usr/sbin/in.telnetd", "-L",
+			 port_helper_bin, NULL };
+
+	/*****************************************/
+
 	new = accept(fd, NULL, 0);
 	if (new < 0)
 		return -errno;
diff -urpN b/arch/um/drivers/ubd_kern.c a/arch/um/drivers/ubd_kern.c
--- b/arch/um/drivers/ubd_kern.c	2013-12-16 19:28:47.770089822 +0100
+++ a/arch/um/drivers/ubd_kern.c	2013-12-16 19:30:49.182093627 +0100
@@ -745,7 +745,7 @@ static void ubd_close_dev(struct ubd *ub
 static int ubd_open_dev(struct ubd *ubd_dev)
 {
 	struct openflags flags;
-	char **back_ptr;
+	char **back_ptr, *tmp;
 	int err, create_cow, *create_ptr;
 	int fd;
 
@@ -754,6 +754,11 @@ static int ubd_open_dev(struct ubd *ubd_
 	create_ptr = (ubd_dev->cow.file != NULL) ? &create_cow : NULL;
 	back_ptr = ubd_dev->no_cow ? NULL : &ubd_dev->cow.file;
 
+
+	tmp = strchr(ubd_dev->file, ',');
+	if (tmp)
+		ubd_dev->file[tmp - ubd_dev->file] = '\0';
+
 	fd = open_ubd_file(ubd_dev->file, &ubd_dev->openflags, ubd_dev->shared,
 				back_ptr, &ubd_dev->cow.bitmap_offset,
 				&ubd_dev->cow.bitmap_len, &ubd_dev->cow.data_offset,
